src/flexible_credential_source.rs

```diff
@@ -0,0 +1,195 @@
+use std::path::{Path, PathBuf};
+
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    authentication_manager::ServiceAccount,
+    custom_service_account::ApplicationCredentials,
+    default_authorized_user::{ConfigDefaultCredentials, UserCredentials},
+    types::HyperClient,
+    CustomServiceAccount, Error,
+};
+
+// Implementation referenced from
+// https://github.com/golang/oauth2/blob/a835fc4358f6852f50c4c5c33fddcd1adade5b0a/google/google.go#L158
+// Currently not implementing external account credentials
+// Currently not implementing impersonating service accounts (coming soon !)
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub(crate) enum FlexibleCredentialSource {
```

Comment:
I think we should ditch the `Flexible` prefix here and just call this `Credentials`, and probably move it to the `types` module.